### ISSUE TITLE
Partykit ceases to function after leaving a tab in the background for a prolonged period

### ISSUE THREAD

**Issue by ahanusek** [partykit/partykit#434]
Hello,

I am currently exploring the capabilities of PartyKit and am testing its resilience with respect to leaving a tab in the background and then returning to the app. Unfortunately, my initial experience has not been positive. Below are the steps to reproduce the issue:

1. Visit the [example reactions page on GitHub](https://github.com/partykit/example-reactions).
2. Open the demo app (in my case with a new ID at [this link](https://example-reactions.jevakallio.partykit.dev/test)).
3. Interact with the reactions; everything works as expected.
4. Leave the tab in the background for a while (approximately 2 minutes).
5. Return to the tab and attempt to add a new reaction.

Upon following these steps, I observed that nothing happens when trying to add a new reaction. Although a new message is sent via WebSocket, the object with reactions is not updated as expected.

<img width="1114" alt="Zrzut ekranu 2023-09-27 o 22 33 56" src="https://github.com/partykit/partykit/assets/16654786/25a59461-f7c6-42e6-9d20-6dd931ab67e5">

https://github.com/partykit/partykit/assets/16654786/80ca7357-d5f6-49b8-9571-84479e41c138

**Comment by threepointone** [comment]
@jevakallio ^ if you’re aware

**Comment by jevakallio** [comment]
Looking at this now!

**Comment by jevakallio** [comment]
@ahanusek thank you for reporting this! I can reproduce the issue -- it seems that the connection routing for hibernatable websockets breaks after a period of inactivity. The connections continue to exist, but the connection attachment metadata is lost, and messages are no longer delivered to the durable object.

This appears to be a recent regression in the Cloudflare runtime. We don't have an automated test case for prolonged background tab behaviour, so it's possible that this issue could have existed prior, but it seems unlikely that we would not have noticed, since we are running dozens of hibernating parties across all of our demos.

I'll create a minimal reproduction and report this to Cloudflare.

Here's a non-hibernating version of the same project, which does not exhibit the same behaviour:
https://example-reactions-without-hibernation.jevakallio.partykit.dev/

**Comment by ahanusek** [comment]
@jevakallio Thanks for the quick answer and explanation.

**Comment by jevakallio** [comment]
I've narrowed down the issue: it seems we lose the websocket attachments when the socket is brought back from hibernation.

In the case of https://example-reactions.jevakallio.partykit.dev, this breaks the connection because the rate limiter depends on serialized socket attachments to keep the rate limiter timestamp. 

Disabling rate limiting on the example fixes the issue there: https://github.com/partykit/example-reactions/commit/5978eb9c5346528a64cdf067566d0b285b58973e

The root cause is still not clear, will continue digging, but will be AFK for a couple of hours now.

**Comment by jevakallio** [comment]
Okay, I found the root cause -- this turned out to be an issue with PartyKit, not Cloudflare.

The short version of the story:
1. In hibernation mode, PartyKit stores basic connection information such as connection id and url in the websocket attachments, so that we can regain access to them when the object wakes up from hibernation.
2. The rate limiter in example-reactions used `serializeAttachment` to keep track of its state.
3. The `serializeAttachment` overwrote the fields internally managed by PartyKit
4. When the object next wakes up in reaction to a message, we assert that the information we store is still present. This assertion failed, and caused a silent error.
5. The error was silent and went unnoticed because of a bug in the Cloudflare tail API (I've previously reported it here https://github.com/cloudflare/workers-sdk/issues/3936)

As a temporary fix, I changed the example code to add to the existing attachment:
https://github.com/partykit/example-reactions/commit/d22ac357b993cc5f955b756ed4a4ba25051b30ea#diff-e5dc3bf6eb0828c0f06d81c0870634da77211acd0505554d93c3e268999e3fccR47

I'll fix this in PartyKit platform code next.
